Kubernetes: Allow generating secured Ingress resources

Sgitario opened the pull request:
Kubernetes exposes applications using https://kubernetes.io/docs/concepts/services-networking/ingress[Ingress resources]. To generate the Ingress resource, just apply the following configuration:

```
quarkus.kubernetes.ingress.expose=true
```

This would generate the following Ingress resource:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    app.quarkus.io/commit-id: a58d2211c86f07a47d4b073ea9ce000d2c6828d5
    app.quarkus.io/build-timestamp: 2022-06-29 - 13:22:41 +0000
  labels:
    app.kubernetes.io/name: kubernetes-with-ingress
    app.kubernetes.io/version: 0.1-SNAPSHOT
  name: kubernetes-with-ingress
spec:
  rules:
    - http:
        paths:
          - backend:
              service:
                name: kubernetes-with-ingress
                port:
                  name: http
            path: /
            pathType: Prefix
```

After deploying these resources to Kubernetes, the Ingress resource will allow unsecured connections to reach out your application. 

To secure the incoming connections, Kubernetes allows enabling https://kubernetes.io/docs/concepts/services-networking/ingress/#tls[TLS] within the Ingress resource by specifying a Secret that contains a TLS private key and certificate. You can generate a secured Ingress resource by simply adding the "tls.secret-name" properties:

```
quarkus.kubernetes.ingress.expose=true
## Ingress TLS configuration:
quarkus.kubernetes.ingress.tls.my-secret.enabled=true
```

This configuration will generate the following secured Ingress resource:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  ...
  name: kubernetes-with-secure-ingress
spec:
  rules:
    ...
  tls:
    - secretName: my-secret
```

Now, Kubernetes will validate all the incoming connections using SSL with the certificates provided within the secret with name "my-secret".

**NOTE:** More information about how to create the secret in https://kubernetes.io/docs/concepts/services-networking/ingress/#tls[here].

Sgitario commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/IngressTlsConfig.java:
```diff
@@ -0,0 +1,23 @@
+     * If true, it will use the TLS configuration in the generated Ingress resource.
+     */
+    @ConfigItem
+    boolean enabled;
```
I don't really like having to add this `enabled` field... However, I didn't see any workaround for https://github.com/quarkusio/quarkus/issues/7862 (this issue is closed, but having optional nested in other optional objects is not working).

iocanel reviewed (approved):
LGTM

Sgitario commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesCommonHelper.java:
```diff
@@ -436,18 +435,6 @@ private static List<DecoratorBuildItem> createAnnotationDecorators(Optional<Proj
                             now.format(DateTimeFormatter.ofPattern("yyyy-MM-dd - HH:mm:ss Z")), new String[0]))));
         }
 
-        if (config.getExposition().isPresent() && config.getExposition().get().expose) {
```
Yep

Sgitario commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/PlatformConfiguration.java:
```diff
@@ -74,10 +74,6 @@ public interface PlatformConfiguration extends EnvVarHolder {
 
     ResourcesConfig getResources();
 
-    default Optional<ExpositionConfig> getExposition() {
```
As this is not a generic code anymore, it's used in specific platforms only.

gsmet commented:
Unfortunately, there is now a conflict. Any chance you could rebase and ping me loudly when done? Thanks!

Sgitario committed:
Kubernetes: Allow generating secured Ingress resources

Kubernetes exposes applications using https://kubernetes.io/docs/concepts/services-networking/ingress[Ingress resources]. To generate the Ingress resource, just apply the following configuration:

[source]
----
quarkus.kubernetes.ingress.expose=true
----

This would generate the following Ingress resource:

[source, yaml]
----
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    app.quarkus.io/commit-id: a58d2211c86f07a47d4b073ea9ce000d2c6828d5
    app.quarkus.io/build-timestamp: 2022-06-29 - 13:22:41 +0000
  labels:
    app.kubernetes.io/name: kubernetes-with-ingress
    app.kubernetes.io/version: 0.1-SNAPSHOT
  name: kubernetes-with-ingress
spec:
  rules:
    - http:
        paths:
          - backend:
              service:
                name: kubernetes-with-ingress
                port:
                  name: http
            path: /
            pathType: Prefix
----

After deploying these resources to Kubernetes, the Ingress resource will allow unsecured connections to reach out your application.

To secure the incoming connections, Kubernetes allows enabling https://kubernetes.io/docs/concepts/services-networking/ingress/#tls[TLS] within the Ingress resource by specifying a Secret that contains a TLS private key and certificate. You can generate a secured Ingress resource by simply adding the "tls.secret-name" properties:

[source]
----
quarkus.kubernetes.ingress.expose=true
quarkus.kubernetes.ingress.tls.my-secret.enabled=true
----

This configuration will generate the following secured Ingress resource:

[source, yaml]
----
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  ...
  name: kubernetes-with-secure-ingress
spec:
  rules:
    ...
  tls:
    - secretName: my-secret
----

Now, Kubernetes will validate all the incoming connections using SSL with the certificates provided within the secret with name "my-secret".

[NOTE]
====
More information about how to create the secret in https://kubernetes.io/docs/concepts/services-networking/ingress/#tls[here].
====

asd

Sgitario commented:
> Unfortunately, there is now a conflict. Any chance you could rebase and ping me loudly when done? Thanks!

PR updated and conflicts resolved!

gsmet commented:
Merged, thank you!
